ast/compile.go

```diff
@@ -1728,6 +1735,218 @@ func (c *Compiler) rewriteDynamicTerms() {
 	}
 }
 
+func (c *Compiler) parseMetadataBlocks() {
+	// Only parse annotations if rego.metadata built-ins are called
+	regoMetadataCalled := false
+	for _, name := range c.sorted {
+		mod := c.Modules[name]
+		WalkRules(mod, func(rule *Rule) bool {
+			WalkExprs(rule, func(expr *Expr) bool {
+				if isRegoMetadataChainCall(expr) || isRegoMetadataRuleCall(expr) {
+					regoMetadataCalled = true
+				}
+				return regoMetadataCalled
+			})
+			return regoMetadataCalled
+		})
+
+		if regoMetadataCalled {
+			break
+		}
+	}
+
+	if regoMetadataCalled {
+		// NOTE: Possible optimization: only parse annotations for modules on the path of rego.metadata-calling module
+		for _, name := range c.sorted {
+			mod := c.Modules[name]
+
+			if len(mod.Annotations) == 0 {
+				var errs Errors
+				mod.Annotations, errs = parseAnnotations(mod.Comments)
+				errs = append(errs, attachAnnotationsNodes(mod)...)
+				for _, err := range errs {
+					c.err(err)
+				}
+			}
+		}
+	}
+}
+
+func (c *Compiler) rewriteRegoMetadataCalls() {
+	for _, name := range c.sorted {
+		mod := c.Modules[name]
+
+		WalkRules(mod, func(rule *Rule) bool {
+			var regoMetadataChainCalled bool
+			var regoMetadataRuleCalled bool
```

Comment:
[nit] the context is clear here, how about shortening these?
Maybe
```suggestion
			chainCalled, ruleCalled := false, false
```